### PR TITLE
Add inference for #917

### PR DESCRIPTION
**Pull request by matwojcik**
Agreed with @kubukoz that this PR will replace his https://github.com/fthomas/refined/pull/919.

I've suggested some naming, made one of inferences more generic, and added scala 3.0+ version.

### CODE REVIEW

File: modules/core/shared/src/test/scala-3.0-/eu/timepit/refined/BooleanInferenceSpec.scala [hunk 1]
```diff
@@ -5,6 +5,8 @@ import eu.timepit.refined.api.Inference
 import eu.timepit.refined.boolean._
```
Comment (kubukoz) [inline modules/core/shared/src/test/scala-3.0-/eu/timepit/refined/BooleanInferenceSpec.scala:5]: I'm kinda worried that this file is only in `scala-3.0-` 🤔

File: modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/boolean.scala [hunk 2]
```diff
@@ -273,6 +273,15 @@ private[refined] trait BooleanInference0 extends BooleanInference1 {
   implicit def conjunctionCommutativity[A, B]: (A And B) ==> (B And A) =
     Inference.alwaysValid("conjunctionCommutativity")
 
+  implicit def substitutionInConjunction[A, B, C](implicit p1: B ==> C): (A And B) ==> (A And C) =
+    p1.adapt("substitutionInConjunction(%s)")
+
+  implicit def disjunctionTautologyElimination[A, B, C](implicit
+      p1: B ==> C,
```
Comment (kubukoz) [inline modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/boolean.scala:280]: why this ordering? does `B ==> C` need to go first?

File: modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/boolean.scala [hunk 3]
```diff
@@ -273,6 +273,15 @@ private[refined] trait BooleanInference0 extends BooleanInference1 {
...
+      p2: A ==> C
+  ): (A Or B) ==> C =
+    Inference.combine(p1, p2, "disjunctionElimination")
```
Comment (kubukoz) [inline modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/boolean.scala:283]: `Combine` is "and", did you mean to make this use "or"? (as we discussed, probably rqeuires a new method)

Comment (matwojcik) [reply modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/boolean.scala:283]: Yep, but we were wrong :) In order to imply that `A Or B` can be replaced by `C`, both `A` and `C` needs to be equivalents of `C`.


Here is the truth table for OR option:

```
  A B C  │  ((B ↔ C) ∨ (A ↔ C)) → ((A ∨ B) ↔ C)
  ───────┼─────────────────────────────────────
  1 1 1  │      1    1    1    *1     1    1   
  1 1 0  │      0    0    0    *1     1    0   
  1 0 1  │      0    1    1    *1     1    1   
  1 0 0  │      1    1    0    *0     1    0   
  0 1 1  │      1    1    0    *1     1    1   
  0 1 0  │      0    1    1    *0     1    0   
  0 0 1  │      0    0    0    *1     0    0   
  0 0 0  │      1    1    1    *1     0    1  
```

File: modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/boolean.scala (hunk 2 above)
Comment (matwojcik) [reply modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/boolean.scala:280]: nope, will swap ;)

**Comment by matwojcik** [comment]
@fthomas could you review this please?

**Review by fthomas — APPROVED** [review]
LGTM. Thanks @matwojcik and @kubukoz!

**Pull request merged**